datafusion/physical-expr/src/utils/guarantee.rs

```diff
@@ -168,14 +195,21 @@ impl LiteralGuarantee {
 
                     // if all terms are 'col <op> literal' with the same column
                     // and operation we can infer any guarantees
+                    //
+                    // For those like (a != bar OR a != baz).
+                    // We can't combine the (a != bar OR a != baz) part, but
+                    // it also doesn't invalidate our knowledge that a !=
+                    // foo is required for the expression to be true.
+                    // So we can only create a multi guarantee for `=`
+                    // (or a single value). (e.g. ignore `a != foo OR a != bar`)
                     let first_term = &terms[0];
                     if terms.iter().all(|term| {
                         term.col.name() == first_term.col.name()
-                            && term.op == first_term.op
+                            && term.guarantee == Guarantee::In
```

Comment:
`a != foo OR a != bar` would be filtered out here, So we don't need to check `new_values.len() == 1` in `aggregate_multi_conjunct`.  If not, `new_values.len() == 1` will confilct with the  LiteralGuarantee result from `InList` who may have multi literals. And I think refactored code is easier to follow for me.

Reply:
I also double checked that this will verify that `first_term.guarantee` needs to be `In` as well ✅